beacon-light-client/plonky2/input_fetchers/balance_verification/deposits_accumulator/lib/pubkey_commitment_mapper_scheduler.ts

```diff
@@ -0,0 +1,230 @@
+  config: SchedulerContextConfig,
+): SchedulerContext {
+  const redis = new Redis(config.redisHost, config.redisPort);
+
+  const ethJsonRPC = new ethers.providers.JsonRpcProvider(
+    config.ethJsonRPCProviderURL,
+  );
+
+  const contract = new Contract(
+    config.contractAddress,
+    config.contractAbi,
+    ethJsonRPC,
+  );
+
+  return {
+    redis,
+    ethJsonRPC,
+    contract,
+    protocol: config.protocol,
+  };
+}
+
+export async function destroySchedulerContext(
+  ctx: SchedulerContext,
+): Promise<void> {
+  await ctx.redis.quit();
+}
+
+export async function pollEvents(
+  ctx: SchedulerContext,
+  timeout: number = 5000,
+): Promise<void> {
+  const lastLoggedBlock = await ctx.redis.client.get(
+    `${ctx.protocol}:${lastLoggedBlockKey}`,
+  );
+  if (lastLoggedBlock === null) {
+    console.log('Error: Tree not initialized, use --rebuild');
+    await destroySchedulerContext(ctx);
+    process.exit(1);
+  }
+
+  while (true) {
+    const lastLoggedBlock = await getLastLoggedBlock(ctx);
+    const headBlockNumber = await ctx.ethJsonRPC.getBlockNumber();
+
+    await fetchEventsAsyncCB(
+      ctx.contract,
+      depositEventName,
+      lastLoggedBlock + 1,
+      headBlockNumber,
+      async event => {
+        const args = event.args!;
+        const pubkey = formatHex(args[0]);
+        await updateDepositState(ctx, pubkey, event.blockNumber);
+      },
+    );
+
+    await setLastLoggedBlock(ctx, headBlockNumber);
+    await sleep(timeout);
+  }
+}
+
+export async function rebuildCommitmentMapperTree(
+  ctx: SchedulerContext,
+): Promise<void> {
+  const contractDeploymentBlockNumber =
+    await queryContractDeploymentBlockNumber(ctx);
+
+  if (contractDeploymentBlockNumber === null) {
+    console.log('Error: Invalid contract address');
+    await destroySchedulerContext(ctx);
+    process.exit(1);
+  }
+
+  await purgePubkeyCommitmentMapperData(ctx);
+
+  await setCurrentlyComputedPubkeyMapping(ctx, 0);
+  await setLastLoggedBlock(ctx, contractDeploymentBlockNumber - 1);
+
+  const headBlockNumber = await ctx.ethJsonRPC.getBlockNumber();
+
+  await fetchEventsAsyncCB(
+    ctx.contract,
+    depositEventName,
+    contractDeploymentBlockNumber,
+    headBlockNumber,
+    async event => {
+      const args = event.args!;
+      const pubkey = formatHex(args[0]);
+      await updateDepositState(ctx, pubkey, event.blockNumber);
+    },
+  );
+
+  await setLastLoggedBlock(ctx, headBlockNumber);
+}
+
+async function updateDepositState(
+  ctx: SchedulerContext,
+  pubkey: string,
+  blockNumber: number,
+): Promise<void> {
+  const pipeline = ctx.redis.client.multi();
+
+  pushPubkeyToProcessingQueuePipe(pipeline, ctx.protocol, pubkey, blockNumber);
+  pushPubkeyToRegistryPipe(pipeline, ctx.protocol, pubkey);
+  setLastLoggedBlockPipe(pipeline, ctx.protocol, blockNumber);
+
+  await pipeline.exec();
+
+  console.log(`[${blockNumber}] ${pubkey}`);
+}
+
+async function queryContractDeploymentBlockNumber(
```

Comment:
Etherscan provider in ethers already provides such functionality mostly.

Reply:
We decided to not depend on etherscan

Reply:
This approach is more portable

Reply:
It is also ran just on rebuild so it doesn't really matter how it's done as long as it works